<a href="https://colab.research.google.com/github/solomeya12345-code/python_for_data_analitic/blob/main/Homework_3_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [11]:
import numpy as np
import pandas as pd

In [7]:
df = pd.read_csv('drive/MyDrive/Python_for_DA/supermarket_sales.csv')

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [34]:
df.head(4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4


In [16]:
df.tail(4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.690,1022.490,3/2/2019,17:16,Ewallet,973.80,4.761905,48.690,4.4
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.592,33.432,2/9/2019,13:22,Cash,31.84,4.761905,1.592,7.7
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.291,69.111,2/22/2019,15:33,Cash,65.82,4.761905,3.291,4.1
999,849-09-3807,A,Yangon,Member,Female,Fashion accessories,88.34,7,30.919,649.299,2/18/2019,13:28,Cash,618.38,4.761905,30.919,6.6


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

Альтернативні відповіді:

In [55]:
display(df[:4], df[-4:])

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4


,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.690,1022.490,3/2/2019,17:16,Ewallet,973.80,4.761905,48.690,4.4
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.592,33.432,2/9/2019,13:22,Cash,31.84,4.761905,1.592,7.7
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.291,69.111,2/22/2019,15:33,Cash,65.82,4.761905,3.291,4.1
999,849-09-3807,A,Yangon,Member,Female,Fashion accessories,88.34,7,30.919,649.299,2/18/2019,13:28,Cash,618.38,4.761905,30.919,6.6


In [56]:
display(df.head(4), df.tail(4))

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4


,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.690,1022.490,3/2/2019,17:16,Ewallet,973.80,4.761905,48.690,4.4
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.592,33.432,2/9/2019,13:22,Cash,31.84,4.761905,1.592,7.7
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.291,69.111,2/22/2019,15:33,Cash,65.82,4.761905,3.291,4.1
999,849-09-3807,A,Yangon,Member,Female,Fashion accessories,88.34,7,30.919,649.299,2/18/2019,13:28,Cash,618.38,4.761905,30.919,6.6


2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [17]:
df[615:620]

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
615,509-29-3912,A,Yangon,Member,Female,Sports and travel,38.72,9,17.4240,365.9040,3/20/2019,12:24,Ewallet,348.48,4.761905,17.4240,4.2
616,327-40-9673,B,Mandalay,Member,Male,Sports and travel,72.60,6,21.7800,457.3800,1/13/2019,19:51,Cash,435.60,4.761905,21.7800,6.9
617,840-19-2096,C,Naypyitaw,Member,Male,Electronic accessories,87.91,5,21.9775,461.5275,3/14/2019,18:10,Ewallet,439.55,4.761905,21.9775,4.4
618,828-46-6863,A,Yangon,Member,Male,Food and beverages,98.53,6,29.5590,620.7390,1/23/2019,11:22,Credit card,591.18,4.761905,29.5590,4.0
619,641-96-3695,C,Naypyitaw,Member,Female,Fashion accessories,43.46,6,13.0380,273.7980,2/7/2019,17:55,Ewallet,260.76,4.761905,13.0380,8.5


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [19]:
df.dtypes

,0
Invoice ID,object
Branch,object
City,object
Customer type,object
Gender,object
Product line,object
Unit price,float64
Quantity,int64
Tax 5%,float64
Total,float64


Альтернативна відповідь:

In [58]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time                     1000 non-null   object 
 12  Payment                  1000 non-null   object 
 13  cogs                     1000 non-null   float64
 14  gross margin percentage  

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [20]:
with pd.option_context("display.precision", 2):
    display(df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99]))

,gross income
count,1000.00
mean,15.38
std,11.71
min,0.51
1%,0.91
25%,5.92
50%,12.09
75%,22.45
99%,45.25
max,49.65


Значення 22.45 відповідає 75-тому персентилю, і це означає, що 75% від усього валового доходу припадає на значення 22.45 або нижче. Я б пояснила, що 75% даних по валовому продкуту є в межах від 0 до 22.45.

In [59]:
df.describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])['gross income']

,gross income
count,1000.000000
mean,15.379369
std,11.708825
min,0.508500
1%,0.913735
25%,5.924875
50%,12.088000
75%,22.445250
99%,45.250750
max,49.650000


  75ому персентилю колонки gross income відповідає значення 22.44 - це означає, що ми заробляємо з 75% товарів суму меншу за 22.44 одиниць.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [22]:
df['City'].value_counts()

,count
City,
Yangon,340
Mandalay,332
Naypyitaw,328


In [60]:
df.City.value_counts()

,count
City,
Yangon,340
Mandalay,332
Naypyitaw,328


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [23]:
df['City'].value_counts(normalize=True)

,proportion
City,
Yangon,0.340
Mandalay,0.332
Naypyitaw,0.328


In [61]:
df.City.value_counts(normalize=True)

,proportion
City,
Yangon,0.340
Mandalay,0.332
Naypyitaw,0.328


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [30]:
df[df['City'].isin(['Yangon', 'Mandalay'])].groupby('City')['gross income'].mean()

,gross income
City,
Mandalay,15.232024
Yangon,14.874001


В Mandalay більше середнє gross income, ніж у Yangon.
Причина не у кількості продажів, адже у Yangon вона більша.

Причина може бути у вартості за одну одиницю, у Mandalay в середньому Unit price вище. Тобто клієнти купляють дорожчі продукти. Можливо через більший рівень доходу в жителів міста. В такому випадку у Yangon потрібно залучати більше покупців.

Також у Mandalay трохи більша середня кількість товарів на одного клієнта.
Відповідно і більший середній чек.

Ще може залежати від статі споживачів (в Mandalay відсоток жінок-покупців переважає Yangon).

Розв'язок:

In [62]:
mean_Yangon = df[df.City=='Yangon']['gross income'].mean()
mean_Mandalay = df[df.City=='Mandalay']['gross income'].mean()

In [63]:
mean_Yangon, mean_Mandalay

(np.float64(14.874001470588235), np.float64(15.232024096385542))

Середнє значення gross income у Yangon становить 14.87, тоді як у Mandalay — 15.23. Це означає, що середній дохід від продажів у супермаркетах у Mandalay трохи вищий, ніж у Yangon. Три гіпотези, чому це може бути так:

1. **Відмінності в споживацькому попиті**: Можливо, у Mandalay вищий попит на деякі товари, що приносять більший прибуток. Це може бути через відмінності в демографії, культурних звичках або доступності альтернативних торгових точок. Аби це зʼясувати - маємо порівняти продажі різних груп товарів в містах і знайти, яка група користується більшим попитом в Mandalay. Можливо варто збільшити поставки цих товарів в цьому місті, аби заробляти більше.

2. **Різниця в пробутковості товарів через локальні правила**: Можливо у різних містах різна прибутковість одних і тих самих товарів, і тому за однакового попиту, середня прибутковість в Mandalay - вища. Аби це зʼясувати можемо дослідити маржу на товари в двох містах.

3. **Різноманітність та доступність асортименту**: У супермаркетах у Mandalay може бути більший або кращий асортимент товарів, що приваблює більше покупців та сприяє збільшенню середнього доходу. Це також може бути пов'язано з ефективнішою маркетинговою стратегією або кращим управлінням запасами. Аби це дослідити в ідеалі мати доступ до товарів, які поставляються в кожному з міст. Можна також переглянути, чи по факту покупок в Mandalay купують більш різноманітні товари або просто інші, ніж в Yangon.

Наведені гіпотези надалі можуть бути підтверджені або спростовані через додатковий аналіз даних, такий як вивчення патернів покупок, аналіз цін та маркетингових активностей, а також оцінка зовнішніх економічних і соціальних факторів у цих містах.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [64]:
margins = df['gross margin percentage'].nunique()
if margins == 1:
  print("У всіх всіх інвойсах однакова маржа")
else:
  print("Маржа відрізняється для різних інвойсів")

У всіх всіх інвойсах однакова маржа


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розподіл кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [54]:
df[df['City'].isin(['Yangon', 'Mandalay'])].groupby('City')['Product line'].value_counts(normalize=True)

City      Product line          
Mandalay  Fashion accessories       0.186747
          Sports and travel         0.186747
          Electronic accessories    0.165663
          Health and beauty         0.159639
          Food and beverages        0.150602
          Home and lifestyle        0.150602
Yangon    Home and lifestyle        0.191176
          Electronic accessories    0.176471
          Sports and travel         0.173529
          Food and beverages        0.170588
          Fashion accessories       0.150000
          Health and beauty         0.138235
Name: proportion, dtype: float64

Розв'язок:

In [65]:
df[df.City=='Yangon']['Product line'].value_counts(normalize=True)

,proportion
Product line,
Home and lifestyle,0.191176
Electronic accessories,0.176471
Sports and travel,0.173529
Food and beverages,0.170588
Fashion accessories,0.150000
Health and beauty,0.138235


In [66]:
df[df.City=='Mandalay']['Product line'].value_counts(normalize=True)

,proportion
Product line,
Fashion accessories,0.186747
Sports and travel,0.186747
Electronic accessories,0.165663
Health and beauty,0.159639
Food and beverages,0.150602
Home and lifestyle,0.150602


In [67]:
df_sel_cols = df[df['City'].isin(['Yangon', 'Mandalay'])]

counts = df_sel_cols.groupby(['City', 'Product line']).size().unstack(fill_value=0)

percentages = counts.div(counts.sum(axis=1), axis=0)

percentages

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
City,,,,,,
Mandalay,0.165663,0.186747,0.150602,0.159639,0.150602,0.186747
Yangon,0.176471,0.150000,0.170588,0.138235,0.191176,0.173529


Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)